main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -1073,44 +1109,50 @@ object Resolver extends Phase[NamedAst.Root, ResolvedAst.Root] {
   }
 
   /**
-    * Finds the def or sig with the qualified name `qname` in the namespace `ns0`.
+    * Finds the def with the namespace `qual` and name `name` in the namespace `ns0`,
+    * Or the sig with the class name `qual` and name `name` in the namespace `ns0`
     */
-  def lookupDefSig(qname: Name.QName, ns0: Name.NName, root: NamedAst.Root): Validation[NameLookupResult, ResolutionError] = {
-    val defOpt = tryLookupDef(qname, ns0, root)
-    val sigOpt = tryLookupSig(qname, ns0, root)
+  def lookupDefOrSig(qual: Name.Ident, name: Name.Ident, ns0: Name.NName, root: NamedAst.Root): Validation[NameLookupResult, ResolutionError] = {
+    // case 1: `qual` is a namespace and `name` is a def
+    val defNamespace = Name.NName(qual.sp1, List(qual), qual.sp2)
+    val defQName = Name.QName(defNamespace.sp1, defNamespace, name, name.sp2)
+    val defOpt = tryLookupDef(defQName, ns0, root)
+
+    // case 2: `qual` is a class and `name` is a sig
+    val classQName = Name.mkQName(qual)
+    val sigOpt = tryLookupSig(classQName, name, ns0, root)
 
     (defOpt, sigOpt) match {
       // Case 1: Can't find the name anywhere
-      case (None, None) => ResolutionError.UndefinedName(qname, ns0, qname.loc).toFailure
+      case (None, None) => ResolutionError.UndefinedName(defQName, ns0, defQName.loc).toFailure
       // Case 2: Can find only a def with the name
       case (Some(defn), None) =>
         if (isDefAccessible(defn, ns0)) {
           NameLookupResult.Def(defn).toSuccess
         } else {
-          ResolutionError.InaccessibleDef(defn.sym, ns0, qname.loc).toFailure
+          ResolutionError.InaccessibleDef(defn.sym, ns0, defQName.loc).toFailure
         }
       // Case 3: Can find only a sig with the name
       case (None, Some(sig)) =>
         if (isSigAccessible(sig, ns0)) {
           NameLookupResult.Sig(sig).toSuccess
         } else {
-          ResolutionError.InaccessibleSig(sig.sym, ns0, qname.loc).toFailure
+          ResolutionError.InaccessibleSig(sig.sym, ns0, sig.loc).toFailure
         }
       // Case 4: Can find both with the name
       case (Some(defn), Some(sig)) =>
         (isDefAccessible(defn, ns0), isSigAccessible(sig, ns0)) match {
           // Case 4.1: Neither is accessible
-          case (false, false) => ResolutionError.InaccessibleDef(defn.sym, ns0, qname.loc).toFailure // MATT need something for if both are inaccessible?
+          case (false, false) => ResolutionError.InaccessibleDef(defn.sym, ns0, defQName.loc).toFailure // MATT need something for if both are inaccessible?
           // Case 4.2: Only the def is accessible
           case (true, false) => NameLookupResult.Def(defn).toSuccess
```

Comment:
I think we need to open (or see if we already have) a ticket with the question:

Should visibility affect scoping? or How should visibility affect scoping?

We currently take visibility into account- and maybe that makes sense- but I am not 100% convinced.

In Java this is the case. No problem in a class having a private method/field and a subclass later defining the same method/field. 

But something about this makes me uneasy.

Thoughts?

Reply:
When can this error even happen? Normally if there's a conflict, we call it at the site of the `use`.

I ran test coverage, and these cases aren't covered (pretty sure that's my fault).

Reply:
This is possible only with top-level classes.

```
namespace N {
    def f(): Int = 123
}
class N {
    def f(): Int
}
...
let a = N.f(); // sig or def?
```
Assuming we report naming conflicts even when visibility would disambiguate, we're partially exposing the implementation which is supposed to be hidden.

But yes, we can pull this into a separate ticket.